### PR TITLE
#7462: Bump react-redux from 7 to 8 and memoize remaining selectors

### LINKED ISSUE #9251: Verify selectors that return fresh array/object references are memoized
Context
---
- Selectors that return fresh array/objects (including literals, or using filter/map) must be cached using `createSelector` to avoid unnecessary re-renders
- Selectors that return a function that construct fresh objects should be memoized with lodash's `memoize`
- See initial discussion: https://github.com/pixiebrix/pixiebrix-extension/pull/9191#discussion_r1790295509

Reference
---
- https://reselect.js.org/api/createSelector
- https://reselect.js.org/api/createStructuredSelector (not available in our version of RTK/reselect, see example of equivalent): https://github.com/pixiebrix/pixiebrix-extension/issues/7462

Progress
---
- [x] modComponentSelectors: 
- [x] modInstanceSelectors 
- [x] editorAnalysisSelectors: selectModComponentAvailability?
- [x] editorClipboardSelectors
- [x] editorDataPanelSelectors
- [x] editorDesignViewSelectors
- [x] editorErrorSelectors: selectErrorState?
- [x] editorModalSelectors: selectEditorModalVisibilities?
- [x] editorModComponentSelectors: selectAllDeletedModComponentIds?
- [x] editorPipelineSelectors
- [x] runtimeSelectors: makeSelectBrickTrace?
- [x] logSelectors
- [x] integrationSelectors
- [x] modsPageSelectors

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Stepping stone to #7462
- Bump react-redux from 7 to 8
- Closes #9251 (react-redux 8 seems to be better about logging console warnings for unmemoized selectors). I opened the Page Editor/Extension Console/Sidebar and looked for console warnings

## Discussion

- `react-redux` 9 (and therefore RTK 2) depends on React 18 because it drops the `useSyncExternalStore` polyfill (and others). We might be able to keep using that shim and bypass the requirement. But IMO we should likely bump to React 18 first

## Remaining Work

- [x] Fix type errors
- [x] Wait for CI E2E tests to run and fix bugs uncovered by E2E tests: infinite re-render, etc.

## Future Work

- Brownbag lightening talk on memoizing selectors

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/components/logViewer/LogCard.tsx [hunk 1]
```diff
@@ -40,9 +41,11 @@ const LogCard: React.FunctionComponent<OwnProps> = ({
   const [page, setPage] = useState(0);
 
   const { isLoading, isError, error } = useSelector(selectLogs);
-  const dispatch = useDispatch();
+  const dispatch = useDispatch<AppDispatch>();
```
[the remark this answers is not in the document]
Comment (grahamlangford) [reply src/components/logViewer/LogCard.tsx:44]: Or something like:
```
type RootStates = OptionsRootState | PageEditorRootState | SidebarRootState;
type SharedRootStates = SharedUnionFieldsDeep<RootStates>;
type SharedDispatch = ThunkDispatch<SharedRootStates, unknown, AnyAction>;
```

File: src/components/logViewer/logSlice.ts [hunk 2]
```diff
@@ -30,6 +32,8 @@ import { castDraft } from "immer";
 
 const REFRESH_INTERVAL = 750;
 
+let timeout: NodeJS.Timeout | null = null;
```
Comment (twschiller) [inline src/components/logViewer/logSlice.ts:35]: Rewrite log polling code to be React 18 safe

File: src/pageEditor/layout/EditorLayout.tsx [hunk 3]
```diff
@@ -63,18 +61,20 @@ const EditorLayout: React.FunctionComponent = () => {
     [isInserting, isRestricted, isStaleSession],
   );
 
-  const previousPane = usePreviousValue(currentPane);
-  if (previousPane !== currentPane) {
+  // Hide modals on Editor state transitions
```
Comment (twschiller) [inline src/pageEditor/layout/EditorLayout.tsx:64]: I'm not sure what caused the behavior change in the react-redux upgrade to trigger this

**Comment by grahamlangford** [comment]
> react-redux 9 (and therefore RTK 2) depends on React 18 because it drops the useSyncExternalStore polyfill (and others). We might be able to keep using that shim and bypass the requirement. But IMO we should likely bump to React 18 first

Yes, react-redux 8 supports React 18. We should definitely upgrade to 18 first

Also, useSelector will now warn us in dev mode if we have a bad selector:
https://github.com/reduxjs/react-redux/releases/tag/v8.1.0

File: src/pageEditor/store/editor/editorSelectors/editorErrorSelectors.ts [hunk 4]
```diff
@@ -17,11 +17,16 @@
 
 import type { EditorRootState } from "@/pageEditor/store/editor/pageEditorTypes";
 import { deserializeError } from "serialize-error";
+import { createSelector } from "@reduxjs/toolkit";
 
-export const selectErrorState = ({ editor }: EditorRootState) => ({
-  isBetaError: editor.error && editor.beta,
-  editorError: editor.error ? deserializeError(editor.error) : null,
-});
+export const selectErrorState = createSelector(
```
[the remark this answers is not in the document]
Comment (grahamlangford) [reply src/pageEditor/store/editor/editorSelectors/editorErrorSelectors.ts:22]: No we weren't. That was introduced in react-redux 8.10 (See my PR comment). The selector runs twice on mount in dev mode and compares the results. Super handy!

[the remark this answers is not in the document]
Comment (grahamlangford) [reply src/pageEditor/store/editor/editorSelectors/editorErrorSelectors.ts:22]: Let's be sure to share that with the team once this is merged

[the remark this answers is not in the document]
Comment (grahamlangford) [reply src/pageEditor/store/editor/editorSelectors/editorErrorSelectors.ts:22]: Things are going to get even better once React 19 and React-compiler are stable. I can't wait to never use `useMemo`, `useCallback`, or `React.memo` again

File: src/sidebar/sidebarSelectors.ts [hunk 5]
```diff
@@ -52,13 +52,23 @@ export const selectSidebarModActivationPanel = ({
...
-  sidebar.modActivationPanel,
-];
+export const selectSidebarEntries = createSelector(
+  [
+    ({ sidebar }: SidebarRootState) => sidebar.panels,
+    ({ sidebar }: SidebarRootState) => sidebar.forms,
+    ({ sidebar }: SidebarRootState) => sidebar.temporaryPanels,
+    ({ sidebar }: SidebarRootState) => sidebar.staticPanels,
+    ({ sidebar }: SidebarRootState) => sidebar.modActivationPanel,
+  ],
```
Comment (grahamlangford) [inline src/sidebar/sidebarSelectors.ts:62]: Any particular reason you chose used an array of selectors? It's perfectly valid of course, just different from how we usually use createSelector

**Review by grahamlangford — APPROVED** [review]

File: src/pageEditor/layout/EditorLayout.tsx (hunk 3 above)
Comment (grahamlangford) [reply src/pageEditor/layout/EditorLayout.tsx:64]: Yes that's a strange one. I wouldn't expect there to be issues there. 🤷

File: src/sidebar/sidebarSelectors.ts (hunk 5 above)
Comment (twschiller) [reply src/sidebar/sidebarSelectors.ts:62]: I was thinking I could avoid the eslint warning on number of function arguments. But that's actually on the arrow function defined, not the call to `createSelector`

They're equivalent, but I'll change before merging for consistency

**Pull request merged**